### DISCUSSION TITLE
[Serializer] Add support for ByteBuffer input/output

### DISCUSSION THREAD

**Discussion by alphaqu** [activej/activej#166] [Ideas]
Allowing for the use of a ByteBuffer opens up a ton of extreme optimizations on the user end. Like using a mapped ByteBuffer to read from a file at lightning speeds, Currently the serializer is very limited to a byte array, which is slow to create in many scenarios.

**Comment by eduard-vasinskyi** [comment]
Hi, @alphaqu
You can already use heap `ByteBuffer`s with `BinarySerializer` by working with `ByteBuffer`’s array directly and manually updating position.

```java
int initialPos = buffer.position();
int newPos = serializer.encode(buffer.array(), initialPos, item);
buffer.position(newPos);
serializer.decode(buffer.array(), initialPos);
```
 
As for mapped `ByteBuffer`s, we do not see how it may improve current performance. Currently, we work directly with byte arrays. With `ByteBuffer`s we would introduce an overhead of additional method invocations. As `BinaryInput`, `BinaryOutput` primarily use operations for primitive types, the overhead of using a `ByteBuffer` would be greater than using byte arrays directly. 

Additionally, if we would use var length encoding, it would translate into N byte-wise operations in a `ByteBuffer`.
If you worry about slow to create arrays, you can use  ActiveJ [bytebuf module](https://activej.io/misc/bytebuf) which comes with a `ByteBuf` pool. Arrays would be created once and then reused per your application needs, mitigating garbage collection overhead. To use `ByteBuf` with a `BinarySerializer` you would also need to operate directly on a byte array, manually modifying buf positions.
